### PR TITLE
kv : Add MvccSnapshot interface

### PR DESCRIPTION
**Pull request by c4pt0r**
MvccSnapshot interface allow user to get/seek *specific* version of data.

### CODE REVIEW

**Comment by c4pt0r** [comment]
@siddontang 
@ngaut 
@qiuyesuifeng 
PTAL

File: kv/kv.go [hunk 1]
```diff
@@ -107,6 +134,8 @@ type Driver interface {
 type Storage interface {
 	// Begin transaction
 	Begin() (Transaction, error)
+	// MvccSnapshot get a snaphot that cloud read any version of data.
```
Comment (shenli) [inline kv/kv.go]: cloud?

File: kv/kv.go [hunk 2]
```diff
@@ -85,6 +102,16 @@ type Transaction interface {
 	LockKeys(keys ...Key) error
 }
 
+// MvccSnapshot is used to get/seek a specific verion in a snaphot.
+type MvccSnapshot interface {
+	// MvccGet returns the specific version of given key, if the version doesn't
+	// exists, returns the nearest(lower) version's data.
+	MvccGet(k Key, ver Version) ([]byte, error)
```
Comment (siddontang) [inline kv/kv.go]: I think Get is enough. The Mvcc prefix seems redundant.
Following NewIterator too.

File: kv/kv.go [hunk 3]
```diff
@@ -85,6 +102,16 @@ type Transaction interface {
...
+	// MvccIterator seeks to the key in the specific version's snapshot, if the
+	// version doesn't exists, return the nearest(lower) version's snaphot.
+	NewMvccIterator(k Key, ver Version) Iterator
```
Comment (shenli) [inline kv/kv.go]: There is a release interface in Snapshot. Does MvccSnapshot need release interface?

File: kv/kv.go (hunk 2 above)
Comment (siddontang) [inline kv/kv.go]: ignore above, I find that dbSnapshot must implement MvccSnapshot and it has Get already.

Comment (c4pt0r) [inline kv/kv.go]: Yep, you got it.

File: kv/kv.go (hunk 3 above)
Comment (c4pt0r) [inline kv/kv.go]: That make sense.

File: kv/kv.go (hunk 1 above)
Comment (c4pt0r) [inline kv/kv.go]: Typo. I'll fix it.

File: kv/kv.go (hunk 3 above)
Comment (c4pt0r) [inline kv/kv.go]: DONE

**Comment by c4pt0r** [comment]
@ngaut 
@siddontang 
PTAL

File: store/localstore/txn.go [hunk 4]
```diff
@@ -225,6 +227,14 @@ func (txn *dbTxn) Commit() error {
 	return txn.doCommit()
 }
 
+func (txn *dbTxn) CommitVersion() (kv.Version, error) {
```
Comment (siddontang) [inline store/localstore/txn.go]: maybe CommittedVersion or other 
Commit is a verb.

File: kv/kv.go [hunk 5]
```diff
@@ -56,12 +60,41 @@ type Version struct {
 	Ver uint64
 }
 
+var (
+	// MaxVersion is the maximum version, notice that it's not a valid version
```
Comment (qiuyesuifeng) [inline kv/kv.go]: Comment end of .

File: kv/kv.go [hunk 6]
```diff
@@ -56,12 +60,41 @@ type Version struct {
...
+	MaxVersion = Version{Ver: math.MaxUint64}
+	// MinVersion is the minimum version, it's not a valid version, too.
```
Comment (qiuyesuifeng) [inline kv/kv.go]: a valid -> an valid

File: kv/kv.go [hunk 7]
```diff
@@ -56,12 +60,41 @@ type Version struct {
...
+	MinVersion = Version{Ver: 0}
+)
+
+// NewVersion is a simple wrapper.
```
Comment (qiuyesuifeng) [inline kv/kv.go]: How about `NewVersion creates a new Version struct.`?

File: kv/kv.go (hunk 6 above)
Comment (qiuyesuifeng) [inline kv/kv.go]: Sorry, it is OK.

File: kv/kv.go [hunk 8]
```diff
@@ -56,12 +60,41 @@ type Version struct {
...
+func NewVersion(v uint64) Version {
+	return Version{
+		Ver: v,
+	}
+}
+
+// Cmp returns the comparison result of two version.
```
Comment (qiuyesuifeng) [inline kv/kv.go]: two versions

File: kv/kv.go [hunk 9]
```diff
@@ -85,6 +121,18 @@ type Transaction interface {
 	LockKeys(keys ...Key) error
 }
 
+// MvccSnapshot is used to get/seek a specific verion in a snaphot.
+type MvccSnapshot interface {
+	// MvccGet returns the specific version of given key, if the version doesn't
+	// exists, returns the nearest(lower) version's data.
+	MvccGet(k Key, ver Version) ([]byte, error)
+	// MvccIterator seeks to the key in the specific version's snapshot, if the
+	// version doesn't exists, return the nearest(lower) version's snaphot.
```
Comment (qiuyesuifeng) [inline kv/kv.go]: doesn't exists -> doesn't exist
return -> returns

Comment (qiuyesuifeng) [inline kv/kv.go]: lower or lowest?

File: store/localstore/txn.go [hunk 10]
```diff
@@ -225,6 +227,14 @@ func (txn *dbTxn) Commit() error {
...
+	// Check if this trx is not committed.
```
Comment (qiuyesuifeng) [inline store/localstore/txn.go]: Check if this txn is commited?

File: store/localstore/mvcc_test.go [hunk 11]
```diff
@@ -133,3 +137,70 @@ func (t *testMvccSuite) TestMvccNext(c *C) {
...
+
+	mvccSnapshot, err := t.s.GetMvccSnapshot()
+	defer mvccSnapshot.MvccRelease()
+	c.Assert(err, IsNil)
+
+	// iter helper function
+	iterFunc := func(it kv.Iterator) bool {
+		found := false
+		for it.Valid() {
+			log.Info(it.Key(), it.Value())
```
Comment (qiuyesuifeng) [inline store/localstore/mvcc_test.go]: Remove this line.

File: store/localstore/mvcc_test.go [hunk 12]
```diff
@@ -133,3 +136,69 @@ func (t *testMvccSuite) TestMvccNext(c *C) {
...
+	// Modify
+	tx, _ = t.s.Begin()
+	err = tx.Set(encodeInt(1), []byte("new"))
+	c.Assert(err, IsNil)
+	err = tx.Commit()
+	c.Assert(err, IsNil)
+	v, err := tx.CommittedVersion()
+	c.Assert(err, IsNil)
+
+	mvccSnapshot, err := t.s.GetMvccSnapshot()
```
Comment (ngaut) [inline store/localstore/mvcc_test.go]: s/GetMvccSnapshot/GetSnapshot/g

Comment (ngaut) [inline store/localstore/mvcc_test.go]: s/mvccSnapshot/snapshot/g

File: store/localstore/txn.go [hunk 13]
```diff
@@ -225,6 +227,14 @@ func (txn *dbTxn) Commit() error {
 	return txn.doCommit()
 }
 
+func (txn *dbTxn) CommittedVersion() (kv.Version, error) {
+	// Check if this trx is not committed.
```
Comment (ngaut) [inline store/localstore/txn.go]: Rename trx to transaction.

File: util/prefix_helper_test.go [hunk 14]
```diff
@@ -133,7 +133,8 @@ func (s *testPrefixSuite) TestPrefix(c *C) {
 	c.Assert(err, IsNil)
 	err = ScanMetaWithPrefix(txn, str, func([]byte, []byte) bool { return true })
```
Comment (ngaut) [inline util/prefix_helper_test.go]: Split into multiple lines.

**Comment by c4pt0r** [comment]
@ngaut 
@qiuyesuifeng 
PTAL

**Comment by qiuyesuifeng** [comment]
LGTM

**Comment by siddontang** [comment]
LGTM

**Pull request merged**